src/components/10-atoms/input-text/index.js

```diff
@@ -0,0 +1,235 @@
+      required: { type: Boolean },
+
+      type: { type: String },
+      disabled: { type: Boolean },
+
+      isReact: { type: Boolean },
+    };
+  }
+
+  static get styles() {
+    return styles;
+  }
+
+  constructor() {
+    super();
+    this.id = '';
+    this.name = '';
+    this.label = '';
+    this.placeholder = '';
+    // text, email, password
+    this.type = 'text';
+    this.error = '';
+    this.validation = false;
+    this.required = false;
+    // experimental
+    this.validation = false;
+    this.valid = true;
+    this.disabled = false;
+    this.isReact = false;
+    this.onFocus = () => {};
+    this.onBlur = () => {};
+    this.onChange = () => {};
+
+    // internal properties
+    this.inputFocus = false;
+    this.wasFocused = false;
+    this.wasBlured = false;
+    this.nativeInput = { value: '' };
+    this.modelValue = '';
+    this.isControlled = false;
+    this.refId = this.id || `input-text-${createRefId()}`;
+  }
+
+  set value(val) {
+    const { isControlled } = this;
+    if (!isControlled && val !== undefined) {
+      this.isControlled = true;
+    }
+    const oldVal = this.modelValue;
+    this.modelValue = val;
+    this.requestUpdate('value', oldVal);
+  }
+
+  get value() {
+    const {
+      isControlled,
+      modelValue,
+      nativeInput: { value: nativeValue },
+    } = this;
+
+    return isControlled ? modelValue : nativeValue;
+  }
+
+  get isRequiredError() {
+    return this.required && !this.value;
+  }
+
+  get isInvalid() {
+    return !this.valid || this.isRequiredError;
+  }
+
+  get showInputError() {
+    return this.isInvalid && this.wasBlured && this.wasFocused;
+  }
+
+  get hideCheckIcon() {
+    if (!this.wasBlured && !this.wasFocused) {
+      return true;
+    }
+
+    if (this.inputFocus) {
+      return true;
+    }
+
+    return this.isInvalid;
+  }
+
+  get hideErrorMessage() {
+    return !this.error || !this.showInputError;
+  }
+
+  get showValidation() {
+    return this.validation || this.required;
+  }
+
+  handleFocus = ev => {
+    this.onFocus(ev);
+    this.inputFocus = true;
+
+    if (!this.wasFocused) {
+      this.wasFocused = true;
+    }
+  };
+
+  handleBlur = ev => {
+    // Validation should be on blur or submit
+    this.onBlur(ev);
+    this.inputFocus = false;
+
+    if (!this.wasBlured) {
+      this.wasBlured = true;
+    }
+  };
+
+  handleInput = ev => {
+    this.onChange(ev);
+
+    // are we a 'controlled' input in the React sense?
+    if (this.isControlled) {
+      // yes, set UI from model state
+      this.nativeInput.value = this.modelValue;
+    }
+  };
+
+  firstUpdated() {
+    this.nativeInput = this.querySelector('input');
+  }
+
+  render() {
+    const {
+      name,
+      required,
+      value,
+      label = '',
+      error = '',
+      type = '',
+      placeholder,
+      disabled,
+      isReact,
+
+      isControlled,
+      refId,
+    } = this;
+
+    this.isControlled = isControlled && isReact;
```

Comment:
It is a bit weird to have a "isControlled" property on the class which will change shortly before rendering and include the value of another property.

Reply:
I started to put the `this.isControlled = isControlled && isReact;` bit into shouldUpdate, which comes before render in the lifecycle. The weirdness is because isReact is another property, which is not available in the value setter. So another way is to explain it via a comment and leave it here. Choose whatever makes more sense to you.